Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/common/calendar/CalendarViewModel.java

```diff
@@ -0,0 +1,256 @@
+/*-
+ * #%L
+ * This file is part of "Apromore Core".
+ * %%
+ * Copyright (C) 2018 - 2020 Apromore Pty Ltd.
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+package org.apromore.portal.common.calendar;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Collections;
+import java.util.Date;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAdjusters;
+import java.time.Instant;
+import java.time.OffsetTime;
+import java.time.OffsetDateTime;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.DayOfWeek;
+import java.text.SimpleDateFormat;
+
+import lombok.Getter;
+import lombok.Setter;
+import org.zkoss.bind.annotation.BindingParam;
+import org.zkoss.bind.annotation.ToClientCommand;
+import org.zkoss.bind.annotation.ToServerCommand;
+import org.zkoss.bind.annotation.Command;
+import org.zkoss.bind.annotation.Init;
+import org.zkoss.bind.annotation.NotifyChange;
+import org.zkoss.bind.annotation.NotifyCommand;
+import org.zkoss.bind.annotation.ExecutionArgParam;
+import org.zkoss.bind.annotation.Init;
+import org.zkoss.bind.BindUtils;
+
+import org.zkoss.zk.ui.util.Clients;
+
+import org.apromore.portal.common.calendar.CalendarItem;
+
+@NotifyCommand(value="doItemListChange", onChange="_vm_.itemList")
+@ToServerCommand({"applyChanges", "dummy", "updateDate", "updateItem", "removeItem", "updateEndTime", "updateStartTime", "updateStartTimeExt"})
+public class CalendarViewModel {
+
+    static final ZoneOffset DEFAULT_ZONE_OFFSET = OffsetDateTime.now().getOffset();
+
+    static final OffsetTime DEFAULT_START_TIME = OffsetTime.of(LocalTime.of(9, 0), DEFAULT_ZONE_OFFSET);
+    static final OffsetTime DEFAULT_END_TIME = OffsetTime.of(LocalTime.of(17, 0), DEFAULT_ZONE_OFFSET);
+
+    static final String WEEKDAY_FORMAT = "EEEE";
+    static final String HOLIDAY_FORMAT = "E, yyyy MMM dd";
+    static final SimpleDateFormat TIME_PARSER = new SimpleDateFormat("HH:mm");
+
```

Comment:
Can we make the calendar UI a separate module (a portal plugin module) instead of adding all new calendar-related packages/code to the portal module? This is to avoid making the Portal module monolithic. 'Calendar Management' doesn't sound like the main responsibility of the portal (the portal only provides access to feature plugins and calendar management is one of them).

Reply:
Just saw a separate module called "Calendar-Portal" below. So we've actually got a separate portal plugin for calendar management, but then I'm confused with some calendar-related code located in the Portal module, and some calendar-related code located in the "Calendar-Portal" module.

Reply:
I am checking this. I think this will be deleted. Will confirm and clean it up.